### PR TITLE
feat: post merge upgrade

### PR DESCRIPTION
**Pull request by 0xmovses**
# Summary
- **RFCs**: [Link to RFC](#./link/to/rfc), [Link to RFC](#./link/to/rfc), or $\emptyset$.
- **Categories**: any of `protocol-units`, `networks`, `scripts`, `util`, `cicd`, or `misc`.

## Description

This PR implements a full on-chain governance flow for the Post-L1-Merge feature flags (notably `DECOMMISSION_CORE_RESOURCES`), adds support for a partial-voting feature, and wires up CI to run the post-merge migration.

- Introduces `propose_post_l1_merge_with_full_governance` and `PROPOSAL_SCRIPTS` for on-chain feature-flag proposals.
- Adds `execution_hash` to the feature-flags release bundle to ensure proposal immutability.
- Updates migration modules and CI configuration to support end-to-end Post-L1-Merge upgrades.

### Changelog

<details>
<summary>Show a summary per file</summary>

| File | Description |
| ---- | ----------- |
| protocol-units/execution/maptos/framework/releases/post-l1-merge/src/vote.rs | Adds governance proposal builder, partial/full voting functions, and tests |
| protocol-units/execution/maptos/framework/releases/post-l1-merge/src/tests/vote.data/enable_partial_governance_voting/sources/main.move | Move script to enable partial governance voting |
| protocol-units/execution/maptos/framework/releases/post-l1-merge/src/tests/vote.data/enable_partial_governance_voting/Move.toml | Defines the Move package for partial voting script |
| protocol-units/execution/maptos/framework/releases/post-l1-merge/src/lib.rs | Registers `cached` and `vote` modules and commit-hash macros |
| protocol-units/execution/maptos/framework/releases/post-l1-merge/src/cached.rs | Sets up MRB release and empty script stub |
| protocol-units/execution/maptos/framework/releases/post-l1-merge/build.rs | Adds `commit_hash_with_script!` for the release tool |
| protocol-units/execution/maptos/framework/releases/post-l1-merge/Cargo.toml | Introduces the post-l1-merge release crate |
| protocol-units/execution/maptos/framework/releases/feature-flags/src/lib.rs | Implements `execution_hash` for feature-flag proposals |
| protocol-units/execution/maptos/framework/migrations/pre-l1-merge-to-post-l1-merge/src/lib.rs | Defines the Post-L1-Merge migration struct and errors |
| protocol-units/execution/maptos/framework/migrations/pre-l1-merge-to-post-l1-merge/src/dot_movement.rs | Integrates the migration into the DotMovement CLI |
| process-compose/movement-full-node/process-compose.test-migrate-post-l1-merge.yml | Adds a process for running the migration test |
| Cargo.toml | Adds `aptos-framework-post-l1-merge-release` to the workspace and updates Aptos dependency revisions |
</details>

# Testing
```
just movement-full-node native build.setup.eth-local.celestia-local.test-migrate-post-l1-merge --keep-tui
```
### 🧠 Overview

This PR introduces a fully on-chain governance flow to propose and activate feature flags — specifically targeting the `DECOMMISSION_CORE_RESOURCES` feature — without relying on the core_resources signer. This replaces the macro-based fast-path that used `core_resources` to directly sign proposals.

### 🧱 Full Governance Flow

The flow implemented in `propose_post_l1_merge_with_full_governance` does the following:

1. **Construct the Feature Flags Proposal**
   - Builds a `SetFeatureFlags<PostL1Merge>` wrapper using the existing release logic.
   - Prepares a `Features` struct with the desired feature flags (e.g., `DECOMMISSION_CORE_RESOURCES`).
   - Generates the compiled bytecode and computes the `execution_hash` using `sha2::Sha256`.

2. **Submit a Governance Proposal**
   - Encodes and submits a `create_proposal_v2` transaction using a validator's stake pool address.
   - Uses the execution hash as the target of the proposal to ensure immutability and validation.

3. **Live Chain Integration**
   - Operates against a real, externally running chain (connected via `rest_client`).
   - Uses a real validator account (`LocalAccount`) from the `maptos_config` to sign and submit the transaction.

### 🗳 Why This Matters

- **No Core Signer Required**: Works even when `core_resources` is inactive, enabling governance upgrades on a live, decentralized network.
- **Auditable**: Uses full governance lifecycle with stake-based voting, eliminating opaque upgrade paths.
- **Composable**: Can be extended to support multi-step proposals, voting, and resolution.

### ✅ Example Usage

```rust
propose_post_l1_merge_with_full_governance(
    &mut validator_account,
    &rest_client,
    chain_id
).await?;
```

## Partial Voting Mechanism test 

This test verifies the behavior of the **partial voting mechanism** in the on-chain governance system.

It simulates the following:

1. **Initial Setup**
   - Creates a test harness and two validator accounts.
   - Stakes tokens for both validators and locks them up (required for voting).

2. **Proposal Creation & Pre-Upgrade Voting**
   - Validator 2 creates a proposal (`proposal_id = 0`).
   - Validator 1 votes on this proposal *before* partial voting is officially enabled.
   - Since partial voting isn't enabled yet, the `voting_power` argument is ignored, and validator 1 uses their full voting power.
   - The test verifies that validator 1 has no voting power left after the vote.

3. **Enable Partial Voting**
   - A special script is executed to enable partial governance voting (as would be done via governance in production).
   - After enabling, the test ensures that **a voter cannot re-vote on a proposal they already voted on before the feature was enabled**.

4. **New Proposal & Edge Case Handling**
   - Validator 1 creates a new proposal (`proposal_id = 1`).
   - The test confirms that voting on a non-existent proposal fails with the correct error.

5. **Partial Voting Behavior**
   - Validator 1 and Validator 2 both have full voting power on the new proposal.
   - Validator 1 votes incrementally:
     - First, uses 100 votes with a "yes" vote.
     - Then, uses 1000 more votes with a "no" vote.
     - The test checks that remaining voting power is reduced correctly after each vote.
   - Finally, validator 1 tries to vote using their full stake, but only the remaining voting power is used.
   - The test ensures a voter cannot exceed their available voting power.

### CODE REVIEW

**Comment by andygolay** [comment]
Looks like some conflicts to resolve. 

Is there an issue for this? I think it should be made clear which requirements for post-L1-merge upgrade this satisfies, and which it doesn't, so we're clear on what progress has been made.

As far as I understand, the requirement for post L1 merge upgrade are:

1. decommission core resources account
2. implement decentralized governance with validators

I see in the PR description, "This PR sets up the testing tool and runs the framework upgrade only." So it sounds like from that, we should infer that both requirements listed above are not met yet. Is that correct? If so, could that be reflected in an issue we can keep track of?

**Comment by 0xmovses** [comment]
Sure I'll write up some issues for the extra bits!

**Comment by 0xmovses** [comment]
@andygolay thing that needs to be done now is 2. full mainnet governance, as we chatted yesterday.

File: networks/movement/movement-full-node/src/admin/bring_up/biarritz_rc1_to_pre_l1_merge/downgrade.rs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+use crate::common_args::MovementArgs;
```
Comment (andygolay) [inline networks/movement/movement-full-node/src/admin/bring_up/biarritz_rc1_to_pre_l1_merge/downgrade.rs:1]: Any idea why the biarritz_rc1_to_pre_l1_merge bring up CLI files showing up as new additions in this diff?

They already exist in `main`: 
image

Comment (andygolay) [reply networks/movement/movement-full-node/src/admin/bring_up/biarritz_rc1_to_pre_l1_merge/downgrade.rs]: Looks like after refreshing this disappeared, so, nevermind :)

File: process-compose/movement-full-node/process-compose.test-migrate-pre-l1-merge-to-post-l1-merge.yml
Comment (andygolay) [inline process-compose/movement-full-node/process-compose.test-migrate-pre-l1-merge-to-post-l1-merge.yml:1]: Is this overlay needed? You're already running `cargo run --bin aptos-framework-post-l1-merge-release-tool` in `process-compose.test-migrate-post-l1-merge.yml`.

Comment (0xmovses) [reply process-compose/movement-full-node/process-compose.test-migrate-pre-l1-merge-to-post-l1-merge.yml:1]: Spuriously added, I'll remove it , thanks

File: protocol-units/execution/maptos/framework/releases/post-l1-merge/mrb_cache/21ec1f2d4367952c103706f51652d47379ce9b1a-post-l1-merge.mrb
Comment (andygolay) [inline protocol-units/execution/maptos/framework/releases/post-l1-merge/mrb_cache/21ec1f2d4367952c103706f51652d47379ce9b1a-post-l1-merge.mrb:1]: This can be deleted, right? Looks like you only need the `867b...` one?

File: protocol-units/execution/maptos/framework/releases/post-l1-merge/mrb_cache/d00f5e5ef3179919b3fc8245ac774f8509ed6a3e-biarritz-rc1.mrb
Comment (andygolay) [inline protocol-units/execution/maptos/framework/releases/post-l1-merge/mrb_cache/d00f5e5ef3179919b3fc8245ac774f8509ed6a3e-biarritz-rc1.mrb:1]: This can be deleted, right? Looks like you only need the `867b...` one?

File: protocol-units/execution/maptos/framework/releases/post-l1-merge/mrb_cache/f3a2758f6e13e4ac3d7e7425c576817358f9b596-post-l1-merge.mrb
Comment (andygolay) [inline protocol-units/execution/maptos/framework/releases/post-l1-merge/mrb_cache/f3a2758f6e13e4ac3d7e7425c576817358f9b596-post-l1-merge.mrb:1]: This can be deleted, right? Looks like you only need the `867b...` one?

Comment (0xmovses) [reply protocol-units/execution/maptos/framework/releases/post-l1-merge/mrb_cache/f3a2758f6e13e4ac3d7e7425c576817358f9b596-post-l1-merge.mrb:1]: I also removed one from the pre-merge package that seemed to be unused.

File: networks/movement/movement-full-node/src/admin/bring_up/biarritz_rc1_to_pre_l1_merge/downgrade.rs (hunk 1 above)
Comment (0xmovses) [reply networks/movement/movement-full-node/src/admin/bring_up/biarritz_rc1_to_pre_l1_merge/downgrade.rs]: It was because I hadn't merged latest main in 😸 , old code was in my PR still

File: protocol-units/execution/maptos/framework/releases/post-l1-merge/src/cached.rs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+	generate_feature_upgrade_module!(feature_upgrade, PostL1Merge, {
+		use aptos_release_builder::components::feature_flags::FeatureFlag;
+		use aptos_types::on_chain_config::FeatureFlag as AptosFeatureFlag;
+
+		// start with the default features and append the Governed Gas Pool feature
+		let mut aptos_feature_flags = AptosFeatureFlag::default_features();
+		// Note: when testing into the future, you may have to use a different revision of [aptos_types] in this crate's Cargo.toml
+		// Or, I suppose you can keep and GOVERNED_GAS_POOL feature flag and a GOVERNED_GAS_POOL_V2 feature flag and just make sure you're disabling the former and enabling the latter. Thereafter, it won't matter what happens to the GOVERNED_GAS_POOL feature flag, i.e., it can be replaced.
+		aptos_feature_flags.push(AptosFeatureFlag::GOVERNED_GAS_POOL);
```
Comment (andygolay) [inline protocol-units/execution/maptos/framework/releases/post-l1-merge/src/cached.rs:65]: Here don't you need to also push `DECOMMISSION_CORE_RESOURCES` to enable that feature flag as part of the release?
image
I get the upgrade to run successfully but as you can see in the screenshot, the flag is not enabled and so the core resources account can still function as such.

**Review by ganymedio — CHANGES_REQUESTED** [review]

File: protocol-units/execution/maptos/framework/releases/post-l1-merge/src/cached.rs (hunk 2 above)
Comment (andygolay) [inline protocol-units/execution/maptos/framework/releases/post-l1-merge/src/cached.rs:65]: Also, we don't need to push `GOVERNED_GAS_POOL` or initialize GGP, right? Since that's already enabled and initialized as of Biarritz RC1?

**Comment by 0xmovses** [comment]
@andygolay Rather than writing loads of small issues, I've just edited the original issue at the start of the sprint with sub-tasks I've complete the ones that have been done. I think this is fine. 

https://github.com/movementlabsxyz/movement-migration/issues/62

File: protocol-units/execution/maptos/framework/releases/post-l1-merge/src/cached.rs (hunk 2 above)
Comment (andygolay) [reply protocol-units/execution/maptos/framework/releases/post-l1-merge/src/cached.rs:65]: @0xmovses I tried doing this locally, I think you may need to pin the latest `aptos-core` rev into Cargo.toml of the `movement` repo to get past the `variant or associated item not found in FeatureFlag` error. 
image

I'll try that now.

Comment (0xmovses) [reply protocol-units/execution/maptos/framework/releases/post-l1-merge/src/cached.rs:65]: Ah I had switched my local change to use movement main branch. I'll change it back and rerun.

Comment (andygolay) [reply protocol-units/execution/maptos/framework/releases/post-l1-merge/src/cached.rs]: image
When I add the feature flag, I get this error 

```
2025-05-07T10:52:58.548513Z  INFO aptos_framework_set_feature_flags_release: Feature flags set
file: "/tmp/nix-shell.gPXtm4/.tmpg3mTJO/proposal.move"
Compiling, may take a little while to download git dependencies...
UPDATING GIT DEPENDENCY https://github.com/movementlabsxyz/aptos-core.git
INCLUDING DEPENDENCY AptosFramework
INCLUDING DEPENDENCY AptosStdlib
INCLUDING DEPENDENCY MoveStdlib
BUILDING proposal
Error: proposing release failed with error: failed to submit gas upgrade proposal: Unknown(Transaction committed on chain, but failed execution: MoveAbort { location: 0000000000000000000000000000000000000000000000000000000000000001::system_addresses, code: 327681, info: Some(AbortInfo { reason_name: "ENOT_CORE_RESOURCE_ADDRESS", description: "The address/account did not correspond to the core resource address" }) })

Caused by:
    failed to submit gas upgrade proposal: Unknown(Transaction committed on chain, but failed execution: MoveAbort { location: 0000000000000000000000000000000000000000000000000000000000000001::system_addresses, code: 327681, info: Some(AbortInfo { reason_name: "ENOT_CORE_RESOURCE_ADDRESS", description: "The address/account did not correspond to the core resource address" }) })
```

Maybe if we set the feature flags last in the release process, that would allow the release to be successfully voted through? But I'm not sure that's possible. Looking into the code now.

Comment (0xmovses) [reply protocol-units/execution/maptos/framework/releases/post-l1-merge/src/cached.rs]: Have you pulled latest from this branch? It's working for me.

Comment (andygolay) [reply protocol-units/execution/maptos/framework/releases/post-l1-merge/src/cached.rs]: One the latest pull I get 
```
Feature flags set
file: "/tmp/nix-shell.gPXtm4/.tmppEehyG/proposal.move"
Compiling, may take a little while to download git dependencies...
UPDATING GIT DEPENDENCY https://github.com/movementlabsxyz/aptos-core.git
INCLUDING DEPENDENCY AptosFramework
INCLUDING DEPENDENCY AptosStdlib
INCLUDING DEPENDENCY MoveStdlib
BUILDING proposal
Error: proposing release failed with error: failed to submit gas upgrade proposal: Unknown(Transaction committed on chain, but failed execution: MoveAbort { location: 0000000000000000000000000000000000000000000000000000000000000001::system_addresses, code: 327681, info: Some(AbortInfo { reason_name: "ENOT_CORE_RESOURCE_ADDRESS", description: "The address/account did not correspond to the core resource address" }) })

Caused by:
    failed to submit gas upgrade proposal: Unknown(Transaction committed on chain, but failed execution: MoveAbort { location: 0000000000000000000000000000000000000000000000000000000000000001::system_addresses, code: 327681, info: Some(AbortInfo { reason_name: "ENOT_CORE_RESOURCE_ADDRESS", description: "The address/account did not correspond to the core resource address" }) })
```    

So, it looks like the feature flag proposal goes through, but the gas upgrade proposal does not. Is that considered a successful outcome? Will we need to execute a gas upgrade proposal for the actual release?

Comment (0xmovses) [reply protocol-units/execution/maptos/framework/releases/post-l1-merge/src/cached.rs]: Do you see these logs:
```

2025-05-07T11:10:42.446214Z  INFO aptos_framework_post_l1_merge_release::cached::full::feature_upgrade: Proposing release PostL1Merge with feature flags
2025-05-07T11:10:42.446234Z  INFO aptos_framework_set_feature_flags_release: Proposing release before feature flags SetFeatureFlags { wrapped_release: PostL1Merge { with_script: RunScript { wrapped_release: PostL1Merge { with_gas_upgrade: GasUpgrade {  ...
```

Then
```
|
UPDATING GIT DEPENDENCY https://github.com/movementlabsxyz/aptos-core.git
INCLUDING DEPENDENCY AptosFramework
INCLUDING DEPENDENCY AptosStdlib
INCLUDING DEPENDENCY MoveStdlib
BUILDING proposal
2025-05-07T11:14:09.287918Z  INFO maptos_framework_release_util: Voted through release
```

I don't see the output you've copied pasted, strange.

Comment (0xmovses) [reply protocol-units/execution/maptos/framework/releases/post-l1-merge/src/cached.rs]: Maybe clear your cache and check the cmd you're running

Comment (andygolay) [reply protocol-units/execution/maptos/framework/releases/post-l1-merge/src/cached.rs]: The command I'm running is
```
just movement-full-node native build.setup.eth-local.celestia-local.test-migrate-post-l1-merge --keep-tui
```

I've done a `cargo clean`, re-running it now. 

You definitely pushed the latest commit?

Comment (andygolay) [reply protocol-units/execution/maptos/framework/releases/post-l1-merge/src/cached.rs]: For `INFO` I get 

image

then

image

then the same error 
image

Comment (0xmovses) [reply protocol-units/execution/maptos/framework/releases/post-l1-merge/src/cached.rs]: I just removed `.movement/` to see if that is the cause.

Comment (0xmovses) [reply protocol-units/execution/maptos/framework/releases/post-l1-merge/src/cached.rs]: > You definitely pushed the latest changes that you made locally?
Yeah no changes to commit locally.

Comment (0xmovses) [reply protocol-units/execution/maptos/framework/releases/post-l1-merge/src/cached.rs]: Ah when I remove .movement/ I see the same error.

**Comment by 0xmovses** [comment]
Yep so I'm pretty sure that yes whilst we set the correct feature flags and is_core_resources fn does indeed return false, this causes an error to throw at somepoint when building and executing the proposal. It may be that the release builder macros generates some code that expected core_resource signer to exist. Let's see I'll be digging in.

**Comment by andygolay** [comment]
> Yep so I'm pretty sure that yes whilst we set the correct feature flags and is_core_resources fn does indeed return false, this causes an error to throw at somepoint when building and executing the proposal. It may be that the release builder macros generates some code that expected core_resource signer to exist. Let's see I'll be digging in.

Yeah technically any proposal after the feature flag update, should fail. So in that sense, it's working as expected. I tried fiddling with the macro wrapping to force the gas upgrade proposal to run first but didn't make much progress. If the gas upgrade isn't needed, we could just remove it and instead run a script to assert that an attempt made by the A550C18 account to implement a governance proposal fails.

**Comment by 0xmovses** [comment]
I think this may actually now be behaving as we want it. I.e core_resource is decommisioned so it cannot pass through releases or upgrades as the signer, it has to be done via governance after that flag is enabled. 

I think the code generated for the macros may be a bit too descriptivie. I'm going to check it out. We wil have to have governance essentially vote in the proposal post core_resource decommsion which is now haha.

**Comment by 0xmovses** [comment]
@andygolay ah sorry still im draft.

**Comment by sebtomba** [comment]
Shouldn't we call [update_governance_config()](https://github.com/movementlabsxyz/aptos-core/blob/main/aptos-move/framework/aptos-framework/sources/aptos_governance.move#L243) with reasonable governance parameters? Or will this occur later?

File: protocol-units/execution/maptos/framework/releases/feature-flags/src/lib.rs [hunk 3]
```diff
@@ -124,6 +125,14 @@ where
 
 		Ok(vec![signed_transaction])
 	}
+
+	pub fn execution_hash(&self) -> Result<Vec<u8>, ReleaseBundleError> {
```
Comment (l-monninger) [inline protocol-units/execution/maptos/framework/releases/feature-flags/src/lib.rs:129]: Within Aptos governance, the means of computing and verifying the `execution_hash` are not quite so straightforward:

- https://github.com/movementlabsxyz/aptos-core/blob/1892d2017f4e31ed5bb18caf49ba282d5c9cbce8/crates/aptos/src/governance/mod.rs#L132
- https://github.com/movementlabsxyz/aptos-core/blob/1892d2017f4e31ed5bb18caf49ba282d5c9cbce8/crates/aptos/src/governance/mod.rs#L1035
- https://github.com/movementlabsxyz/aptos-core/blob/1892d2017f4e31ed5bb18caf49ba282d5c9cbce8/aptos-move/aptos-release-builder/src/components/mod.rs#L760
